tree-building/tree_test.go

```diff
@@ -0,0 +1,313 @@
+package tree
```

Comment:
Consider making this
```suggestion
package tree_test
```

which will enforce that you only call exported functions from `package tree` which is _generally_ but not always the right thing (occasionally you'll have a compelling reason to test an unexported func). This means that your calls to `Build` would become `tree.Build`.